### PR TITLE
Add option to export files by content type

### PR DESCRIPTION
**Pull request by felixjung**
Hi,

For our Jekyll site at @sumup we were looking for a way to import and include content from Contentful. One requirement is that we need to be able to easily create one page per entry while using different layouts based on content type. We found this handy little Jekyll plugin: [avillafiorita/jekyll-datapage_gen](https://github.com/avillafiorita/jekyll-datapage_gen).

However, it expects each data type (i.e. a blog posts, products, job ads) to be included in its own YAML/JSON file. We thought this might in general be a nice option for storing our content from Contentful inside the Jekyll project.

Looking at your nice plugin we couldn't find a way to split up the Contentful data. I implemented the feature and would love to see it get merged. This is close to all the Ruby I've written in my life and I haven't fully updated the tests, yet. What do you think?

Can you actually help me mock the collection of entries with different content types? Never done any of this before and I bet you'd be able to give me a quick lift?

*TODO*
- [ ] Incorporate feedback.
- [ ] Group files from one space within a folder.
- [ ] Update tests to check for one vs. multiple export files.

### CODE REVIEW

**Comment by felixjung** [comment]
I guess one could also change the paths so that content type files are grouped within a folder representing their space? Thus the export paths would look more like `_data/contentful/spaces/{{space_name}}/{{contentType}}.yml`?

File: lib/jekyll-contentful-data-import/data_exporter.rb [hunk 1]
```diff
@@ -17,9 +17,24 @@ def initialize(name, entries, config = {})
 
       def run
         setup_directory
+        puts config
+        export_fn = config['multiple_files'] ? export_by_content_type : export
+        export_fn
```
Comment (dlitvakb) [inline lib/jekyll-contentful-data-import/data_exporter.rb]: `export_fn =` is not needed, as well as the following line, as it would already return the value of either `#export_by_content_type` or `#export`

File: lib/jekyll-contentful-data-import/data_exporter.rb [hunk 2]
```diff
@@ -17,9 +17,24 @@ def initialize(name, entries, config = {})
...
+      end
 
+      def export
+        grouped_entries = ::Jekyll::Contentful::Serializer.new(entries, config).serialize
```
Comment (dlitvakb) [inline lib/jekyll-contentful-data-import/data_exporter.rb]: This line is repeated in both export methods, therefore would be nice if it's moved to `#run` and pass the result as a parameter

File: lib/jekyll-contentful-data-import/data_exporter.rb [hunk 3]
```diff
@@ -34,8 +49,8 @@ def destination_directory
         File.join(base_directory, DATA_FOLDER, CONTENTFUL_FOLDER, SPACES_FOLDER)
       end
 
-      def destination_file
-        File.join(destination_directory, "#{name}.yaml")
+      def destination_file(file_name = name)
```
Comment (dlitvakb) [inline lib/jekyll-contentful-data-import/data_exporter.rb]: You could change this to have the default as `path = ["#{name}.yaml"]` and then on the `File.join...` change it to `File.join(*[destination_directory].concat(path))` that way you could also have subpaths for content types with individual files if desired.

This will require a change on the `#destination_file` call to look like:

```ruby
grouped_entries.each do |content_type, entry_list|
  entry_list.each do |entry|
    File.open(destination_file([content_type, "#{entry[:id]}.yaml"]), 'w') do |file|
      file.write(::Jekyll::Contentful::Serializer.new(entries, config).serialize)
    end
  end
end
```

Or something along those lines

File: spec/jekyll-contentful/data_exporter_spec.rb [hunk 4]
```diff
@@ -75,7 +75,7 @@
 
       it 'serializes entries onto data file' do
         allow(File).to receive(:open).and_yield(StringIO.new)
-        expect_any_instance_of(::Jekyll::Contentful::Serializer).to receive(:to_yaml)
+        expect(YAML).to receive(:dump)
```
Comment (dlitvakb) [inline spec/jekyll-contentful/data_exporter_spec.rb]: Raw `YAML` serialization is not the same as what the serializer does.

Comment (felixjung) [inline spec/jekyll-contentful/data_exporter_spec.rb]: Uhm yeah, this is probably me never having written Ruby tests. I understood the original test would test that the exporter calls the serializer's `to_yaml` method. Since it's no longer doing that, I was testing that it calls `YAML.dump`. Probably I should add a test that it also calls the `serialize` method of the Serializer?

**Comment by felixjung** [comment]
Thanks for the feedback! I'll take care of it next week.

File: spec/jekyll-contentful/data_exporter_spec.rb (hunk 4 above)
Comment (dlitvakb) [inline spec/jekyll-contentful/data_exporter_spec.rb]: Hey @felixjung,

It was ok in your context, so no worries about that, the thing is that the Contentful Serializer doesn't do the same as a simple `YAML.dump`, as it runs through either the default Mapper, or through user defined Mappers, which transform the resulting YAML however the Mapper is set.

Therefore, removing the Serializer is something we don't want

**Comment by felixjung** [comment]
I've taken another look at this today and noticed a couple of things. Having also inspected the middleman plugin, I actually don't see much of a difference between the two current versions. Both will return one file per configured `space`. Or am I missing anything that lead to your original comment @dlitvakb?

What I missed in my original assessment of the plugin is that each configured `space` can actually be interpreted as a query, rather than a space. I can achieve what I was looking for with a configuration like this:

```yaml
contentful:
  spaces:
    - foo:
        space: cfexampleapi
        access_token: b4c0n73n7fu1
        cda_query:
          content_type: 'foo'
    - bar:
        space: cfexampleapi
        access_token: b4c0n73n7fu1
        cda_query:
          content_type: 'bar'
```

One could argue that not having to repeat the `space` and `access_token` might be an improvement (i.e. introducing some global versions). However, after experimenting with various alternative configuration structures, I find that other structures have drawbacks like heavy nesting once you want to introduce a maximum amount of flexibility.

So from my perspective I feel like you could close this PR as I find it unnecessary. I would be interested in two other points though:

1. Can we rename the configuration key `spaces` to `queries`?
2. Can we make the destination folder structure more flexible? Currently it appears that `dest/_data/contentful/spaces/*.yaml` is hard coded. I don't see why the spaces directory is necessary, given that each configured space/query has its own identifier. As far as I can tell, nothing will ever get generated inside the `contentful` folder besides the `spaces` folder itself. So I would argue it would help to add a `dest` property in addition to `base_path`.

What do you think?

**Comment by dlitvakb** [comment]
Hey @felixjung,

The multi-space configuration for same spaces, is something that I personally use a lot when I have multiple content types within the same space that represent different contexts, so, yes, that's a very good approach, and one that I personally really like.

As for how the Middleman plugin works, its by creating 1 folder per space alias, which contains, 1 folder per declared content type, and within each content type folder, 1 file for each entry, indexed by entry ID.

So it's similar to the approach I was proposing in the requested changes, and an approach that might be really useful for people using `data-gen` plugins for Jekyll, as it expects 1 yaml file per mapped page, and this would fit that model.

Regarding being able to remove the `contentful.space` keys from the path, I'm fine with that, already commented in your PR, only a change of the key would be needed and adding it to the documentation.

Regarding changing `spaces` to `queries`, I don't agree with that change, as `queries` is a concept that's "alien" to the core of Contentful, while having an alias for a `space` is part of the core concepts.

Summing up,

* Overriding `destination` folder (but has to be optional) 👍 - it's a handy shortcut, can be used for `data-gen` plugins
* One file per entry, indexed by Content Types (but has to be optional) 👍 - `data-gen` users will be happy
* Changing `spaces` to `queries` 👎 - breaks compatibility and is not a core concept
* All new changes introduced must be optional, we don't want to break existing implementations of this integration

Cheers

**Comment by felixjung** [comment]
Thanks for the feedback. Given that I'm half-way there, I'll finish this one. I did not know about the data-gen part and clearly was not going for the one-file-per-entry approach. But it's sounds cool 👍
